cachi2/core/models/output.py

```diff
@@ -0,0 +1,81 @@
+from pathlib import Path
+from typing import Literal, Optional
+
+import pydantic
+
+from cachi2.core.models.validators import unique_sorted
+
+# Supported package types (a superset of the supported package *manager* types)
+PackageType = Literal["gomod", "go-package"]
+
+
+class Dependency(pydantic.BaseModel):
+    """Metadata about a resolved dependency."""
+
+    type: PackageType
+    name: str
+    version: Optional[str]  # go-package stdlib dependencies are allowed not to have versions
+
+    @pydantic.validator("version")
+    def check_version_vs_type(cls, version: Optional[str], values: dict) -> Optional[str]:
+        """Check that the dependency has a version or is 'go-package'."""
+        ptype = values.get("type")
+        if ptype is not None and (version is None and ptype != "go-package"):
+            raise TypeError(f"{values['type']} dependencies must have a version")
+        return version
+
+
+class Package(pydantic.BaseModel):
+    """Metadata about a resolved package and its dependencies."""
+
+    type: PackageType
+    path: Path  # relative from source directory
+    name: str
+    version: str
+    dependencies: list[Dependency]
+
+    @pydantic.validator("path")
+    def check_path(cls, path: Path) -> Path:
+        """Check that the package path is relative."""
+        if path.is_absolute():
+            raise ValueError(f"package path must be relative: {path}")
+        return path
+
+    @pydantic.validator("dependencies")
+    def unique_deps(cls, dependencies: list[Dependency]) -> list[Dependency]:
+        """Sort and de-duplicate dependencies."""
+        return unique_sorted(dependencies, by=lambda dep: (dep.type, dep.name, dep.version))
+
+
+class EnvironmentVariable(pydantic.BaseModel):
+    """An environment variable."""
+
+    name: str
+    value: str
+    kind: Literal["literal", "path"]
+
+
+class RequestOutput(pydantic.BaseModel):
+    """Results of processing one or more package managers."""
+
+    packages: list[Package]
+    environment_variables: list[EnvironmentVariable]
+
+    @pydantic.validator("packages")
+    def unique_packages(cls, packages: list[Package]) -> list[Package]:
+        """Sort packages and check that there are no duplicates."""
+        return unique_sorted(
+            packages,
+            by=lambda pkg: (pkg.type, pkg.name, pkg.version),
```

Comment:
Should we add this key to a @property in the Package and Depedency classes?

Reply:
I think I had it like that at some point, but then I moved the key to the place where the unique-ification and sorting actually happens. Dependencies are unique in the context of a Package and Packages are unique in the context of the whole Request, so I figured I would let those decide "how unique". Plus the key is used for sorting as well, which felt less obvious when it was defined on the classes.